src/Zcu/PerThread.zig

```diff
@@ -2795,21 +2795,32 @@ pub fn getErrorValueFromSlice(pt: Zcu.PerThread, name: []const u8) Allocator.Err
 /// Removes any entry from `Zcu.failed_files` associated with `file`. Acquires `Compilation.mutex` as needed.
 /// `file.zir` must be unchanged from the last update, as it is used to determine if there is such an entry.
 fn lockAndClearFileCompileError(pt: Zcu.PerThread, file: *Zcu.File) void {
-    switch (file.getMode()) {
-        .zig => {
-            const zir = file.zir orelse return;
-            if (!zir.hasCompileErrors()) return;
-        },
-        .zon => {
-            const zoir = file.zoir orelse return;
-            if (!zoir.hasCompileErrors()) return;
+    const maybe_has_error = switch (file.status) {
+        .never_loaded => false,
+        .retryable_failure => true,
+        .astgen_failure => true,
+        .success => switch (file.getMode()) {
+            .zig => has_error: {
+                const zir = file.zir orelse break :has_error false;
+                break :has_error zir.hasCompileErrors();
+            },
+            .zon => has_error: {
+                const zoir = file.zoir orelse break :has_error false;
+                break :has_error zoir.hasCompileErrors();
+            },
         },
+    };
+
+    // If runtime safety is on, let's quickly lock the mutex and check anyway.
+    if (!maybe_has_error and !std.debug.runtime_safety) {
```

Comment:
`std.debug.runtime_safety` has a deprecation comment on it: https://github.com/ziglang/zig/blob/3a4bb47fedbb890dc149622e31c75101b14c3b16/lib/std/debug.zig#L162-L168

Reply:
that deprecation notice may become nullified by https://github.com/ziglang/zig/issues/22285

Reply:
not it doesn't

> This proposal does **not** suggest removing per-module optimization modes.

Reply:
This is true, but I'm intentionally going to leave it for now because this bit of the language design is still a bit shaky.